lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+            add_line_string(geom)
+          when Feature::Polygon
+            add_polygon(geom)
+          when Feature::GeometryCollection
+            add_collection(geom)
+          end
+        end
+
+        # Adds a LineString or LinearRing
+        # to the graph.
+        #
+        # @param geom [RGeo::Feature::LineString]
+        def add_line_string(geom)
+          # TODO: should we handle empty linestrings?
+          add_edges(geom.segments)
+
+          hedge = unless geom.is_empty?
+                    @incident_edges[geom.start_point.coordinates].first
+                  end
+
+          @geom_edges << GeomEdge.new(hedge, nil)
+        end
+
+        # Adds a Polygon to the graph.
+        #
+        # Will add each shell seperately and find a CCW half-edge
+        # for the exterior ring and a CW half-edge for interior rings
+        # since these are designated as being on the interior of the polygon.
+        #
+        # It is possible that the half-edge will be nil if a suitable one cannot
+        # be found. This indicates the polygon is likely invalid.
+        #
+        # @param geom [RGeo::Feature::Polygon]
+        def add_polygon(geom)
+          exterior = geom.exterior_ring
+          add_edges(exterior.segments)
+
+          hedge = find_hedge(exterior)
+
+          interior_hedges = []
+          geom.interior_rings.each do |interior|
+            add_edges(interior.segments)
+            interior_hedges << find_hedge(interior, ccw: false)
+          end
+
+          @geom_edges << GeomEdge.new(hedge, interior_hedges)
+        end
+
+        # Adds a GeometryCollection to the graph.
+        #
+        # @param col [RGeo::Feature::GeometryCollection]
+        def add_collection(col)
+          col.each do |geom|
+            add_geometry(geom)
+          end
+        end
+
+        # Finds a Half-Edge that is part of a CCW or CW rotation
+        # from the input ring. Returns nil if none found.
+        #
+        # Will only consider half-edges that are colinear with
+        # the first or last segments of the ring.
+        #
+        # @param ring [RGeo::Feature::LinearRing]
+        # @param ccw [Boolean] true for CCW, false for CW
+        # @return [HalfEdge, nil]
+        def find_hedge(ring, ccw: true)
+          return nil if ring.num_points.zero?
+          ccw_target = ccw ? 1 : -1
+
+          coords = ring.start_point.coordinates
+          hedges = incident_edges[coords]
+
+          # find half-edges that are colinear to the start or end
+          # segment of the ring.
+          start_seg = Segment.new(ring.start_point, ring.point_n(1))
+          end_seg = Segment.new(ring.point_n(ring.num_points - 2), ring.end_point)
+          colinear_hedges = hedges.select do |he|
+            start_seg.side(he.destination).zero? || end_seg.side(he.destination).zero?
+          end
+
+          colinear_hedges.each do |hedge|
+            pts = []
+            hedge.each do |he|
```

Comment:
Here comes the cycling issue you were talking about then.

I guess it is important not to get stuck in a infinite loop. Either the Set will be enough, or we need to find something else.

And, you're right, `each` doesn't feel correct there.